File: tools/shot/build_errors.py
"""Summarise a ninja log into error classes rather than error lines.

A `ninja -k 0` run over this tree produces a log where each FAILED entry is
followed by the full clang-cl command line -- roughly 6 KB of flags per failure.
Reading the raw log costs more than the information in it. What actually drives
the next edit is: which diagnostic, how many times, and in which files.

Groups by the normalised diagnostic text (identifiers, types and numbers inside
quotes are kept, since 'use of undeclared identifier X' is a different problem
for each X), and prints the file list for each.

Usage:
  build_errors.py <log> [--limit N] [--files]
"""

import re
import sys
from collections import defaultdict

# clang-cl: `../../path/file.cc(122,10): error: text`
ERROR = re.compile(r"^(\S+?)\((\d+),\d+\): (?:error|fatal error): (.*)$")
# link errors: `lld-link: error: text`
LINK = re.compile(r"^(?:lld-link|ninja): (?:error|fatal error): (.*)$")


def main():
    if len(sys.argv) < 2:
        sys.exit(__doc__)
    limit = 40
    if "--limit" in sys.argv:
        limit = int(sys.argv[sys.argv.index("--limit") + 1])
    show_files = "--files" in sys.argv

    classes = defaultdict(list)
    failed = 0
    with open(sys.argv[1], encoding="utf-8", errors="replace") as f:
        for line in f:
            line = line.rstrip("\n")
            if line.startswith("FAILED:"):
                failed += 1
                continue
            m = ERROR.match(line)
            if m:
                classes[m.group(3)].append("%s:%s" % (m.group(1), m.group(2)))
                continue
            m = LINK.match(line)
            if m:
                classes[m.group(1)].append("<link>")

    print("%d FAILED edge(s), %d distinct diagnostic(s)"
          % (failed, len(classes)))
    for text in sorted(classes, key=lambda t: -len(classes[t]))[:limit]:
        sites = classes[text]
        print("%4d  %s" % (len(sites), text))
        if show_files:
            for s in sorted(set(sites))[:8]:
                print("        %s" % s)
            if len(set(sites)) > 8:
                print("        ... %d more" % (len(set(sites)) - 8))


main()

File: scripts/build-errors.ts
// Summarise a ninja log into error classes rather than error lines.
//
// Reading the raw log costs more than the information in it. What actually
// drives the next edit is: which diagnostic, how many times, and in which
// files. build-engine prints the class view after every build through
// lib/ninja-log.ts; this is the same view on demand, plus the per-edge one.
//
//   pnpm build:errors out/Shot/build.log                 # classes, commonest first
//   pnpm build:errors out/Shot/build.log --limit 10 --files
//   pnpm build:errors out/Shot/build.log --targets       # one entry per failed edge
//   pnpm build:errors out/Shot/build.log --targets --full
//   pnpm build:errors out/Shot/build.log --top 30        # kinds, quoted parts collapsed
//
// A relative log path is resolved against the repository root.

import {readFileSync} from 'node:fs';

import {cac} from 'cac';

import {formatClasses, formatFailures, formatKinds} from './lib/ninja-log.ts';
import {resolve} from './lib/repo.ts';

const cli = cac('build-errors');
cli.command('<log>', 'reduce a ninja log to what went wrong')
    .option('--limit <n>', 'classes to print', {default: 40})
    .option('--files', 'list the files under each class')
    .option('--targets', 'one entry per failed edge, then a count per target')
    .option('--full', 'with --targets: every diagnostic line, not just the first')
    .option('--top <n>', 'group failures by kind and print the commonest')
    .action((logArg: string, options: {limit: number; files?: boolean; targets?: boolean; full?: boolean; top?: number}) => {
      const log = readFileSync(resolve(logArg), 'utf8');
      if (options.top) {
        const out = formatKinds(log, Number(options.top));
        console.log(out || `no failures in ${logArg}`);
      } else if (options.targets) {
        const out = formatFailures(log, {full: options.full === true});
        console.log(out || `no failures in ${logArg}`);
      } else {
        console.log(formatClasses(log, {limit: Number(options.limit), files: options.files === true}));
      }
    });
cli.help();
cli.parse();
